### DISCUSSION TITLE
Weird interaction with JDK21 method Reflection and mocks

### DISCUSSION THREAD

**Discussion by ekfliu** [mockito/mockito#3345] [Q&A]
Hi, I have included a quick maven project to figure out a weird issues with jdk21.
This same exact test work with jdk 17 but fails with jdk 21.

I am unsure of what the problem is, is it an issues with Mockito or JDK 21 itself?
I am expecting InvocationTargetException when doing method reflection invokes, but when a mock is force to throw NullPointerException the JDK21 is throwing IllegalArgumentException? 

This only happens with NullPointerException other exception works fine???
Concrete class throwing null pointer also have no issues.

Here is the quick snip of test code in project

```java
@Slf4j
public class Jdk21ReflectMockitoTest {
	private RuntimeException error = new NullPointerException("Unit Test");
//	private RuntimeException error = new IllegalStateException("Unit Test");

	@Test
	@SneakyThrows
	public void testMockMethod() {
		TestService service = Mockito.mock(TestService.class);
		Mockito.doThrow(error).when(service).testMethod1(Mockito.anyString());
		Method method = TestService.class.getMethod("testMethod1", String.class);
		try {
			Object result = method.invoke(service, "param1");
			Assertions.fail("exception expected");
		} catch (Throwable t) {                    
			log.warn("exception expected", t);
			Assertions.assertTrue(t instanceof InvocationTargetException);
			//fails here! exception is IllegalArgumentException!
		}
	}

	@Test
	@SneakyThrows
	public void testActualMethod() {
		TestService service = new TestServiceImpl("ActualClass", error);
		Method method = TestService.class.getMethod("testMethod1", String.class);
		try {
			Object result = method.invoke(service, "param1");
			Assertions.fail("exception expected");
		} catch (Throwable t) {
			log.warn("exception expected", t);
			Assertions.assertTrue(t instanceof InvocationTargetException);
		}
	}

	public interface TestService {
		String testMethod1(String param);
	}

	public static class TestServiceImpl implements TestService {
		private String prefix;
		private RuntimeException error;

		public TestServiceImpl(String prefix, RuntimeException error) {
			super();
			this.prefix = prefix;
			this.error = error;
		}

		@Override
		public String testMethod1(String param) {
			if (error != null) {
				throw error;
			}
			return prefix + ".testMethod1(" + param + ")";
		}

	}
}
```

[MockitoJdk21Test.zip](https://github.com/mockito/mockito/files/15336743/MockitoJdk21Test.zip)

**Comment by yuvkun10** [comment]
This is a JDK 21 behavior change, not a Mockito bug. The root cause is [JEP 358: Helpful NullPointerExceptions](https://openjdk.org/jeps/358), combined with changes to how Mockito's ByteBuddy-generated proxies interact with JDK 21's more aggressive NPE handling.

**What's happening:**

In JDK 21, when a `NullPointerException` is thrown from within a method invoked via reflection, the JVM's enhanced NPE machinery in certain circumstances wraps or reclassifies the exception differently compared to JDK 17. Specifically, when Mockito's proxy intercepts the call and throws the NPE, it goes through a different code path in the JDK 21 bytecode than a concrete class would.

The `IllegalArgumentException` you're seeing is actually: **"argument type mismatch"** — it's the JDK's reflection `Method.invoke` choking on the way Mockito's proxy is responding to the call, because ByteBuddy-generated proxies in older Mockito versions have a subtle incompatibility with JDK 21's method handle changes.

**The fix:**

Upgrade Mockito to **5.x** (which fully supports JDK 21):

```xml
<dependency>
    <groupId>org.mockito</groupId>
    <artifactId>mockito-core</artifactId>
    <version>5.11.0</version>
    <scope>test</scope>
</dependency>
```

Mockito 5.x switched to using inline mock making by default and updated its ByteBuddy dependency to one that properly handles JDK 21's reflection changes.

**If you must stay on Mockito 4.x**, add the inline mock maker:

```xml
<dependency>
    <groupId>org.mockito</groupId>
    <artifactId>mockito-inline</artifactId>
    <version>4.11.0</version>
    <scope>test</scope>
</dependency>
```

This uses a JVM agent-based approach that bypasses the proxy issue entirely.

**Why only NullPointerException?** JDK 21's NPE enhancement specifically intercepts NPEs for the "helpful NPE message" feature, which triggers a different code path in the proxy interception than other exception types — that's why `IllegalStateException` works fine but `NullPointerException` doesn't.

Hope this helps! Let me know if you need clarification.
